src/main/java/networkbook/model/person/Graduation.java

```diff
@@ -0,0 +1,235 @@
+        if (group < 1 || group > 3) {
+            return null;
+        }
+        String normalizedGradString = gradString.trim().toUpperCase();
+        Pattern pattern = Pattern.compile(AY_REGEX);
+        Matcher matcher = pattern.matcher(normalizedGradString);
+        if (!matcher.matches()) {
+            return null;
+        }
+        String match = matcher.group(group);
+        return match;
+    }
+
+    /**
+     * Parses user input {@code gradString} into an int representing the first calendar year of the academic year.
+     *
+     * @param gradString provided in user command
+     * @return corresponding first calendar year
+     */
+    public static int parseAcadYearStart(String gradString) {
+        String match = matchGroup(gradString, 1);
+        int shortYear = Integer.parseInt(match);
+        int addedPart = (shortYear < YEAR_THRESHOLD) ? YEAR_PART_HI : YEAR_PART_LO;
+        return Integer.parseInt(match) + addedPart;
+    }
+
+
+    /**
+     * Parses user input {@code gradString} into an int representing the second calendar year of the academic year.
+     *
+     * @param gradString provided in user command
+     * @return corresponding second calendar year
+     */
+    public static int parseAcadYearEnd(String gradString) {
+        String match = matchGroup(gradString, 2);
+        int shortYear = Integer.parseInt(match);
+        int addedPart = (shortYear < YEAR_THRESHOLD) ? YEAR_PART_HI : YEAR_PART_LO;
+        return Integer.parseInt(match) + addedPart;
+    }
+
+    /**
+     * Parses user input {@code gradString} into a {@code Semester}.
+     *
+     * @param gradString provided in user command
+     * @return corresponding semester
+     */
+    public static Semester parseSemester(String gradString) {
+        String match = matchGroup(gradString, 3);
+        if (match == null) {
+            return Semester.INVALID;
+        }
+        switch (match) {
+        case "1":
+            return Semester.S1;
+        case "2":
+            return Semester.S2;
+        default:
+            return Semester.INVALID;
+        }
+    }
+
+    /**
+     * Returns true if a given string representation of
+     * a graduation year and semester is valid.
+     */
+    public static boolean isValidGraduation(String gradString) {
+        String normalizedGradString = gradString.trim().toUpperCase();
+        if (normalizedGradString.matches(AY_REGEX)) {
+            int year1 = parseAcadYearStart(gradString);
+            int year2 = parseAcadYearEnd(gradString);
+            Semester sem = parseSemester(gradString);
+            return isValidGraduationYear(year1, year2) && isValidGraduationSemester(sem);
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Returns true if a pair of calendar years is valid (consecutive).
+     */
+    public static boolean isValidGraduationYear(int year1, int year2) {
+        return year2 - year1 == 1;
```

Comment:
Very good implementation. To nitpick a bit, is it true that the input "AY6970-S1" will not be a valid graduation year? I think it would be better to deal with this edge case separately.

Reply:
Hi yes, AY6970 will not be a valid graduation year. I opted not to deal with this edge case separately as I thought it would introduce ambiguity between 2069/2070 and 1969/1970. Do you have any ideas on how best to deal with this issue?

Reply:
My main concern here is that 6970 not being treated as valid input may seem like a bug instead of a feature. I suggest you can just treat it as 2069/2070 and change the command format to be 1970 - 2070 inclusive

Reply:
Ok got it! I will change the user guide to mention the range of AYs instead to avoid further confusion. Thanks :)